builder/builder.go

```diff
@@ -0,0 +1,390 @@
+	keys, pattern []any
+}
+
+func (v *variant) String() string {
+	var s string
+
+	for i := range v.keys {
+		switch v := v.keys[i].(type) {
+		case string:
+			s += v + " "
+		case int, int8, int16, int32, int64, uint, uint8, uint16, uint32, uint64:
+			s += fmt.Sprintf("%d ", v)
+		case float32, float64:
+			s += fmt.Sprintf("%f ", v)
+		default:
+			panic(fmt.Sprintf("unsupported key type: %T", v))
+		}
+	}
+
+	s += "{{"
+
+	for i := range v.pattern {
+		if i != 0 {
+			s += " "
+		}
+
+		switch v := v.pattern[i].(type) {
+		case string:
+			s += v
+		case *Expression:
+			s += v.String()
+		default:
+			panic(fmt.Sprintf("unsupported pattern type: %T", v))
+		}
+	}
+
+	return s + "}}"
+}
+
+type local struct {
+	expr     *Expression
+	variable variable
+}
+
+type literal any
+
+type variable string
+
+type function struct {
+	name    string
+	options []option
+}
+
+type Expression struct {
+	spacing   string
+	operand   any // literal or variable
+	functions []function
+}
+
+func (e *Expression) String() string {
+	s := "{" + e.spacing
+
+	switch v := e.operand.(type) {
+	case variable:
+		s += string(v)
+	case literal:
+		s += printLiteral(v)
+	default:
+		panic(fmt.Sprintf("unsupported operand type: %T", v))
+	}
+
+	for _, f := range e.functions {
+		s += " " + f.name
+
+		for _, o := range f.options {
+			s += " " + o.key + e.spacing + "=" + e.spacing + fmt.Sprint(o.operand)
+		}
+	}
+
+	return s + e.spacing + "}"
+}
+
+func Expr() *Expression { return new(Expression) }
+
+func (e *Expression) Literal(v any) *Expression {
+	e.operand = v
+	return e
+}
+
+func (e *Expression) Var(v string) *Expression {
+	if len(v) == 0 {
+		panic("variable name cannot be empty")
+	}
+
+	if v[0] != '$' {
+		panic(fmt.Sprintf("variable must start with $: %s", v))
+	}
+
+	e.operand = variable(v)
+
+	return e
+}
+
+type option struct { //nolint:govet
+	key     string
+	operand any // literal or variable
+}
+
+func Option(key string, operand any) option { return option{key, operand} }
+
+func (e *Expression) Func(name string, option ...option) *Expression {
+	if len(name) == 0 {
+		panic("function name cannot be empty")
+	}
+
+	switch name[0] {
+	default:
+		panic(fmt.Sprintf("function MUST start with :, + or -: %s", name))
+	case ':', '+', '-':
+		e.functions = append(e.functions, function{name: name, options: option})
+		return e
+	}
+}
+
+func printLiteral(l any) string {
+	switch v := l.(type) {
+	case string:
+		v = escapeChars(v, map[rune]struct{}{
+			'\\': {},
+			'|':  {},
+		})
+
+		return "|" + v + "|"
+	case int, int8, int16, int32, int64, uint, uint8, uint16, uint32, uint64:
+		return fmt.Sprintf("|%d|", v)
+	case float32, float64:
+		return fmt.Sprintf("|%f|", v)
+	default:
+		panic(fmt.Sprintf("unsupported literal type: %T", v))
+	}
+}
+
+// helpers
+
+// specialChars is a list of characters that need to be escaped in the node.Text,
+// because they are either syntax characters or reserved keywords.
+var specialChars = map[rune]struct{}{
```

Comment:
Can we please stop using maps for static list checking in Golang? Please use `switch` instead. 

Also, please read @VladislavsPerkanuks comment on special characters. This is old stuff.